From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rong Bao <webmaster@csmantle.top>
Date: Sun, 4 Oct 2026 01:50:12 +0000
Subject: [PATCH 1/6] Bug 2077820 - [riscv64] Use `tail` instead of `j` for
 tail-calling SharedStub. r=#xpcom-reviewers!

This practice is already adopted by some distros, like Arch Linux RISC-V
and openRuyi.

Link: https://github.com/felixonmars/archriscv-packages/blob/2662cf75438febc6ea993b5ecd6a2c747e8942bb/firefox/xptcstubs-riscv64-tail.patch
Link: https://github.com/openRuyi-Project/openRuyi/blob/ef71a3c657919f7fdffe5ce5c7782abbf8d7549a/SPECS/firefox/2000-riscv64-Use-long-tail-jump-for-xptcall-stubs.patch

Differential Revision: https://phabricator.services.mozilla.com/D330581
---
 xpcom/reflect/xptcall/md/unix/xptcstubs_riscv64.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xpcom/reflect/xptcall/md/unix/xptcstubs_riscv64.cpp b/xpcom/reflect/xptcall/md/unix/xptcstubs_riscv64.cpp
index 17c0789eff48d3e1836a735f1055721b2257865b..3330b34e8da6e0cd318a2fb3d7769a4f14933eed 100644
--- a/xpcom/reflect/xptcall/md/unix/xptcstubs_riscv64.cpp
+++ b/xpcom/reflect/xptcall/md/unix/xptcstubs_riscv64.cpp
@@ -139,7 +139,7 @@ extern "C" nsresult ATTRIBUTE_USED PrepareAndDispatch(nsXPTCStubBase* self,
       ".err   \"stub number "#n" >= 1000 not yet supported\"\n"                     \
       ".endif \n\t"                                                                 \
       "li      t0, "#n" \n\t"                                                       \
-      "j       SharedStub \n"                                                       \
+      "tail    SharedStub \n"                                                       \
       ".if "#n" < 10\n\t"                                                           \
       ".size   _ZN14nsXPTCStubBase5Stub"#n"Ev,.-_ZN14nsXPTCStubBase5Stub"#n"Ev\n\t" \
       ".elseif "#n" < 100\n\t"                                                      \

base-commit: 2150cb1a1339be3581f1a238dcc4a787bceba448
-- 
2.55.0

